### DISCUSSION TITLE
Incremental Static Regenerating Sitemap would be exciting!

### DISCUSSION THREAD

**Discussion by HunterJoey** [vercel/next.js#18671] [Ideas]
**Only if** we can let the page response with a file,
then  /pages/sitemap.xml.js  would response with sitemap.xml.
Imagine that , a sitemap utilizing incremental static regeneration=>You can revalidate the page every X seconds to update the sitemap, and it will be static-blazing-fast and always up to date , **even if you have a 50Mb sitemap**.

So , can we response a page with file?

**Comment by SamiSousa** [comment]
I'm curious about this too, since with ISR the pages on my site would be changing whenever content gets added to or removed from the CMS.

**Comment by PaulSenon** [comment]
will be really useful indeed !
All the logic already exists, if we could just chose to serve a raw xml file without using all the react app html context, it would allow us to take advantage of all this powerful stuff ! 
Or maybe we could think forward and add this incremental static response logic to /api routes ???
So we will be able to have instant response on some api routes that can justify an incremental static response. And therefore we won't call the lambda at each call.
I think this way we will be able to cleanly serve any generated files, of any format, with this super technology.
What do you think about that ?

**Reply by PaulSenon** [reply]
After diving a bit more into that, it seems like it is already possible using vercel cdn through cache control directives. The cached data will not be stored statically on the origin server as for incremental static pages, but it mimic the same behaviour on hedge servers.

just have to set the http cache-control header with:
- a maxAge
- the swr directive

like so:
`cache-control: s-maxage=60, stale-while-revalidate`

It will result in a cache of 60 sec on your endpoint, with SWR (meaning that it will serve the outdated data while fetching the new ones in background)

useful links:
- https://vercel.com/docs/configuration#project/headers
- https://vercel.com/docs/edge-network/caching#stale-while-revalidate

I still think it could be nice to have this feature at the origin so we don't have to rely on a cdn. I just found this workaround that I assume it works for my usecase. Didn't tested yet though...

**Reply by HunterJoey** [reply]
> After diving a bit more into that, it seems like it is already possible using vercel cdn through cache control directives. The cached data will not be stored statically on the origin server as for incremental static pages, but it mimic the same behaviour on hedge servers.
> 
> just have to set the http cache-control header with:
> 
> * a maxAge
> * the swr directive
> 
> like so:
> `cache-control: s-maxage=60, stale-while-revalidate`
> 
> It will result in a cache of 60 sec on your endpoint, with SWR (meaning that it will serve the outdated data while fetching the new ones in background)
> 
> useful links:
> 
> * https://vercel.com/docs/configuration#project/headers
> * https://vercel.com/docs/edge-network/caching#stale-while-revalidate
> 
> I still think it could be nice to have this feature at the origin so we don't have to rely on a cdn. I just found this workaround that I assume it works for my usecase. Didn't tested yet though...

That is doable, thanks for your work. File serving is too clumsy with current solution.

**Comment by MichaelBurgess** [comment]
Does anyone have any examples of this? I have a large (~1,500 pages) next.js SSG site that is set to increase in size a lot in the coming weeks and the build is not the quickest right now, so I want to look into incremental static regeneration as a likely solution to add new SSG pages on demand.

I'm curious about the sitemap though, as I currently use `next-sitemap` and have a `postbuild` script that calls this: `"postbuild": "next-sitemap"`. However, this relies on all the static pages being there at the time of the next.js build, but I'll need a way to regenerate this periodically, or preferably adopt a more ISG-like approach and update the sitemap as I go along.

How are you handling sitemaps in your solutions that use incremental static regeneration? (_if this is a little diff to original qu I'm happy to start a new thread_)

**Reply by BrunoQuaresma** [reply]
I'm having the same issue.

**Reply by rserafim** [reply]
I have a project with over 6 million pages. I need to add or remove (if a customer joins or leaves.)

Besides, I need the sitemap to already exist, even without having the static page generated. For Google to recognize and index faster.

**Reply by metawise** [reply]
Did you guys resolve this? I'm facing similar situation.
